### PR TITLE
[feature] Add Region feature configs and extraction file

### PR DESCRIPTION
**Pull request by vedanuj**
This PR adds the following components :

- Base configs for region features : C4, DC5, FPN
- Model specific config for X-152 backbone
- Script for extracting region features
- Helper functions in `roi_heads` for getting rois features from `AttributeStandardROIHeads` and `AttributeRes5ROIHeads`
- Data loader `build_detection_test_loader_for_images` for building a test dataset from an image folder
- `TestDatasetMapper` class for mapping test datasets with only images

### CODE REVIEW

File: configs/Base-RCNN-region-c4.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+MODEL:
+  META_ARCHITECTURE: "GeneralizedRCNN"
+  ATTRIBUTE_ON: True
+  RPN:
+    PRE_NMS_TOPK_TEST: 6000
+    POST_NMS_TOPK_TEST: 1000
+    SMOOTH_L1_BETA: 0.1111
```
Comment (endernewton) [inline configs/Base-RCNN-region-c4.yaml:7]: Could you add this fix: https://github.com/facebookresearch/grid-feats-vqa/commit/a941ed5577eca23d72dab9cc1e032a62b52003d0#diff-83def4a609ba11f377175844aa343eddR12 to the file? Similarly for other yaml files

File: configs/Base-RCNN-region-dc5.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  META_ARCHITECTURE: "GeneralizedRCNN"
+  ATTRIBUTE_ON: True
+  RESNETS:
+    OUT_FEATURES: ["res5"]
+    RES5_DILATION: 2
+  RPN:
+    IN_FEATURES: ["res5"]
+    PRE_NMS_TOPK_TEST: 6000
+    POST_NMS_TOPK_TEST: 1000
+    SMOOTH_L1_BETA: 0.1111
```
Comment (endernewton) [inline configs/Base-RCNN-region-dc5.yaml:11]: Same for here.

File: configs/Base-RCNN-region-fpn.yaml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  ANCHOR_GENERATOR:
+    SIZES: [[32], [64], [128], [256], [512]]  # One size for each in feature map
+    ASPECT_RATIOS: [[0.5, 1.0, 2.0]]  # Three aspect ratios (same for all in feature maps)
+  RPN:
+    IN_FEATURES: ["p2", "p3", "p4", "p5", "p6"]
+    PRE_NMS_TOPK_TRAIN: 2000  # Per FPN level
+    PRE_NMS_TOPK_TEST: 1000
+    POST_NMS_TOPK_TRAIN: 2000
+    POST_NMS_TOPK_TEST: 1000
+    SMOOTH_L1_BETA: 0.1111
```
Comment (endernewton) [inline configs/Base-RCNN-region-fpn.yaml:19]: Same for here.

File: configs/X-152-region-fpn.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+_BASE_: "Base-RCNN-updn.yaml"
```
Comment (endernewton) [inline configs/X-152-region-fpn.yaml]: I think this line should -fpn.yaml

File: extract_feature_region.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
+#!/usr/bin/env python3
```
Comment (endernewton) [inline extract_feature_region.py]: maybe this file can be renamed 'extract_region_feature'. I can rename the other file 'extract_grid_feature'

File: extract_feature_region.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    add_attribute_config,
+    build_detection_test_loader_with_attributes,
+    build_detection_test_loader_for_images,
+)
+
+# A simple mapper from object detection dataset to VQA dataset names
+dataset_to_folder_mapper = {}
+dataset_to_folder_mapper["coco_2014_train"] = "train2014"
+dataset_to_folder_mapper["coco_2014_val"] = "val2014"
+dataset_to_folder_mapper["visual_genome"] = "visual_genome"
```
Comment (endernewton) [inline extract_feature_region.py]: now we need to double check if visual_genome set is correspondent to the VG features used for VQA. Right now this is only for object/attribute detection.

File: extract_feature_region.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        with torch.no_grad():
+            image_id = inputs[0]["image_id"]
+
+            # compute features and proposals
+            images = model.preprocess_image(inputs)
+            features = model.backbone(images.tensor)
+            proposals, _ = model.proposal_generator(images, features)
+
+            # pooled features and box predictions
+            box_features, pooled_features = model.roi_heads.get_rois_features(
```
Comment (endernewton) [inline extract_feature_region.py]: nit pick: get_rois_features to get_roi_features? happy to discuss more about it.

File: extract_feature_region.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+            cls_probs = F.softmax(predictions[0], dim=-1)
+            cls_probs = cls_probs[:, :-1]  # background is last
+            predictions, r_indices = model.roi_heads.box_predictor.inference(
+                predictions, proposals
+            )
+
+            # postprocess
+            height = inputs[0].get("height")
+            width = inputs[0].get("width")
+            r = postprocessing.detector_postprocess(predictions[0], height, width)
```
Comment (endernewton) [inline extract_feature_region.py]: I remember there is some tiny detail in getting the region features exactly right in the original code base, I remember Aman and myself have discussed before. Is it exactly following the setting of previously extracted region features?

Comment (endernewton) [reply extract_feature_region.py]: Please discuss with Aman about this, being able to exactly reproduce the previous procedure to extract box features is most important thing for this PR.

File: extract_feature_region.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+            info = {
+                "bbox": bboxes.cpu().numpy(),
+                "num_boxes": bboxes.size(0),
+                "objects": classes.cpu().numpy(),
+                "image_height": r.image_size[0],
+                "image_width": r.image_size[1],
+                "cls_prob": cls_probs.cpu().numpy(),
+            }
+            features = pooled_features.cpu().numpy()
+            np.save(os.path.join(dump_folder, str(image_id) + "_info.npy"), info)
```
Comment (endernewton) [inline extract_feature_region.py]: this is now region features are saved? I am not sure storing information separately is a good idea, as the feature folder already has a ton of files, don't want to double the length of the list.

File: extract_feature_region.py (hunk 7 above)
Comment (endernewton) [reply extract_feature_region.py]: Also `box_features` seem not used?

File: configs/Base-RCNN-region-c4.yaml [hunk 10]
```diff
@@ -0,0 +1,30 @@
+MODEL:
```
Comment (endernewton) [inline configs/Base-RCNN-region-c4.yaml:1]: I think we can release some pre-trained model vs region features for people to use as well. And also need to update the README.md. Could you train the models (e.g. C4, DC5, FPN one each), and extract the features? I will update the readme myself.

File: configs/Base-RCNN-region-fpn.yaml [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+  ROI_HEADS:
+    NAME: "AttributeStandardROIHeads"
+    IN_FEATURES: ["p2", "p3", "p4", "p5"]
+    NUM_CLASSES: 1600
+  ROI_BOX_HEAD:
+    NAME: "FastRCNNConvFCHead"
+    NUM_CONV: 4
```
Comment (endernewton) [inline configs/Base-RCNN-region-fpn.yaml]: why using 4conv1fc head here with FPN? why not just use the 2fc layer?

File: configs/Base-RCNN-region-fpn.yaml [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    NUM_FC: 1
+    POOLER_RESOLUTION: 7
+    POOLER_SAMPLING_RATIO: 2
+    SMOOTH_L1_BETA: 1.
+    CLS_AGNOSTIC_BBOX_REG: True
```
Comment (endernewton) [inline configs/Base-RCNN-region-fpn.yaml]: why cls-agnostic?

**Review by endernewton — CHANGES_REQUESTED** [review]

File: extract_feature_region.py (hunk 8 above)
Comment (vedanuj) [reply extract_feature_region.py]: I haven't seen any major difference in the performance yet compared to old features. However I will check if we can get the features before the box head regression and if it helps the performance.

File: extract_feature_region.py (hunk 9 above)
Comment (vedanuj) [reply extract_feature_region.py]: We are saving in this manner as this is the old format of input in Pythia (mmf). I have no strong preference here and I can keep it to one file. However we will need another extraction script that we can keep in mmf repository in that case.

File: configs/Base-RCNN-region-c4.yaml (hunk 10 above)
Comment (vedanuj) [reply configs/Base-RCNN-region-c4.yaml:1]: I have trained the models. I have to extract features for the dc5 and fpn variants.

File: extract_feature_region.py (hunk 8 above)
Comment (vedanuj) [reply extract_feature_region.py]: cc @apsdehal Do you know what was the difference in extracting features before and after box regression?

Comment (endernewton) [reply extract_feature_region.py]: I just did a bit digging and think the original one uses the proposal boxes directly before second regression. https://github.com/facebookresearch/mmf/blob/4517e13f8407f95840340e90a6d30df344d08ea9/tools/scripts/features/extract_features_vmb.py#L184

I think this is more than observed performance difference -- if there is any subtle difference we are aware of that can cause a difference, it should be stated at least. And past experience suggest that having the additional box regression does not help, and can sometimes hurt the accuracy. So for a faithful implementation (ideally it should be able to extract the original BUTD features), I still prefer for it to be implemented just as what is being done before.

File: extract_feature_region.py (hunk 9 above)
Comment (endernewton) [reply extract_feature_region.py]: Since you said old, does it mean in the new format you guys are considering to merge it into one file? I wonder the advantage of having separate files -- those meta information, if not saved with each individual feature, should be saved in a single file (since the size is small) like .json/.npy to save flooding inodes.

Comment (vedanuj) [reply extract_feature_region.py]: I have merged these to a single file.

Comment (endernewton) [reply extract_feature_region.py]: Thanks! If you have time, please fix other comments -- especially the ones of extracting region features exactly following the original BUTD procedure.

File: extract_feature_region.py (hunk 8 above)
Comment (vedanuj) [reply extract_feature_region.py]: I have added ability to extract features from both last FC (fc7) and second last FC (fc6) layers for FPN and DC5 models. Also now we use proposal bbox instead of predicted boxes exactly similar to old pythia implementation. 

I am still keeping ability to have last layer features because current BERT based models don't finetune last layer and hence should use last layer FC features.

**Comment by vedanuj** [comment]
@endernewton This is ready for review again. Uploaded all models, metrics and features to AWS. Readme is updated as well.

File: extract_region_feature.py [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+            proposals, _ = model.proposal_generator(images, features)
+
+            # pooled features and box predictions
+            _, pooled_features, pooled_features_fc6 = model.roi_heads.get_roi_features(
+                features, proposals
+            )
+            predictions = model.roi_heads.box_predictor(pooled_features)
+            cls_probs = F.softmax(predictions[0], dim=-1)
+            cls_probs = cls_probs[:, :-1]  # background is last
+            predictions, r_indices = model.roi_heads.box_predictor.inference(
```
Comment (endernewton) [inline extract_region_feature.py:87]: Are duplicating r_indices removed in this case? iirc it is removed in the BUTD code. Duplicate means the same proposal box corresponding to two different objects.

File: README.md [hunk 14]
```diff
@@ -70,6 +107,14 @@ We release several pre-trained models for grid features: one with R-50 backbone,
...
 | X-152++  | 3.7 | <a href="https://dl.fbaipublicfiles.com/grid-feats-vqa/X-152pp/X-152pp.pth">model</a>&nbsp;\| &nbsp;<a href="https://dl.fbaipublicfiles.com/grid-feats-vqa/X-152pp/metrics.json">metrics</a>&nbsp;\| &nbsp;<a href="https://dl.fbaipublicfiles.com/grid-feats-vqa/X-152pp/X-152pp-features.tgz">features</a> |
 
+We release pretrained models for region features: X-152 with C4, X-152 with DC5, X-152 with FPN:
+
+| Backbone | AP<sub>50:95</sub> | Download |
+| -------- | ---- | -------- |
+| X-152-region-FPN    | 5.25 | <a href="https://dl.fbaipublicfiles.com/grid-feats-vqa/region-fpn-X-152/region-fpn-X-152.pth">model</a>&nbsp;\| &nbsp;<a href="https://dl.fbaipublicfiles.com/grid-feats-vqa/region-fpn-X-152/fpn-X-152-metrics.json">metrics</a>&nbsp;\| &nbsp;<a href="https://dl.fbaipublicfiles.com/grid-feats-vqa/region-fpn-X-152/region-fpn-X-152-features_fc7.tar.gz">features (fc7)</a> &nbsp;<a href="https://dl.fbaipublicfiles.com/grid-feats-vqa/region-fpn-X-152/region-fpn-X-152-features.tar.gz">features (fc6)</a> |
```
Comment (endernewton) [inline README.md:114]: nitpick: maybe just showing one digit after decimal point is enough here.

File: configs/Base-RCNN-region-c4.yaml [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
```
Comment (endernewton) [inline configs/Base-RCNN-region-c4.yaml:1]: For consistency we can either use "region" for all the files, or "updn" in all the files.

File: extract_region_feature.py [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+            info = {
+                "bbox": bboxes.cpu().numpy(),
+                "num_boxes": bboxes.size(0),
+                "objects": classes.cpu().numpy(),
+                "image_height": r.image_size[0],
+                "image_width": r.image_size[1],
+                "cls_prob": cls_probs.cpu().numpy(),
+                "features": pooled_features.cpu().numpy()
+            }
+            np.save(os.path.join(dump_folder, str(image_id) + ".npy"), info)
```
Comment (endernewton) [inline extract_region_feature.py:128]: other parts lgtm, ideally we should test if given the same pre-trained model, using the previous code in mb-benchmark and the current code in D2 should generate the same results. :)

File: extract_region_feature.py [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+        help="name of the dataset",
+        default="coco_2014_train",
+        choices=["coco_2014_train", "coco_2014_val", "coco_2015_test"],
+    )
+    parser.add_argument(
+        "--dataset-path",
+        help="path to image folder dataset, if not provided expects a detection dataset",
+        default="",
+    )
+    parser.add_argument(
```
Comment (endernewton) [inline extract_region_feature.py:55]: maybe worth adding an option of "K" which stands for extracting the top K region features, I think this is useful for a lot of papers that choose top 36, or varying size from 36-100 (maybe another option is needed there)

File: grid_feats/roi_heads.py [hunk 18]
```diff
@@ -226,12 +270,19 @@ def _forward_box(self, features, proposals):
...
         assert len(self.in_features) == 1
 
         features = [features[f] for f in self.in_features]
         return features[0]
 
+    def get_roi_features(self, features, proposals):
+        features = [features[f] for f in self.in_features]
+        box_features = self.box_pooler(features, [x.proposal_boxes for x in proposals])
+        fc7, fc6  = self.box_head(box_features)
+        return box_features, fc7, fc6
```
Comment (endernewton) [inline grid_feats/roi_heads.py:288]: I am not sure if it is trivial for users or not, but if we are storing fc6 anyway, might be worth saving the fc7 weights at the same time so that it can be directly loaded in mmf.

**Review by endernewton — CHANGES_REQUESTED** [review]
Overall looks great! Added a few comments and we are getting there 👍

**Comment by nilinykh** [comment]
Hello,
I was curious if you are planning to adjust the code for storing attributes and their scores as well.
As far as I can see, at the moment the code does not output attribute probabilities and attributes themselves.

Also,
> maybe worth adding an option of "K" which stands for extracting the top K region features, I think this is useful for a lot of papers that choose top 36, or varying size from 36-100 (maybe another option is needed there)

is there a quick way to set the K number of extracted regions to the desired parameter? In particular, 36, as it says in the comment.

Thank you.
